server/src/main/java/org/elasticsearch/search/aggregations/support/BuiltinValuesSourceType.java

```diff
@@ -0,0 +1,214 @@
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.IndexGeoPointFieldData;
+import org.elasticsearch.index.fielddata.IndexNumericFieldData;
+import org.elasticsearch.index.fielddata.IndexOrdinalsFieldData;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.RangeFieldMapper;
+import org.elasticsearch.script.AggregationScript;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.AggregationExecutionException;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.function.LongSupplier;
+
+public enum BuiltinValuesSourceType implements Writeable, ValuesSourceType {
+    ANY {
+        @Override
+        public ValuesSource getEmpty() {
+            // TODO: Implement this or get rid of ANY
+            throw new UnsupportedOperationException("ValuesSourceType.ANY is still a special case");
+        }
+
+        @Override
+        public ValuesSource getScript(AggregationScript.LeafFactory script, ValueType scriptValueType) {
+            // TODO: Implement this or get rid of ANY
+            throw new UnsupportedOperationException("ValuesSourceType.ANY is still a special case");
+        }
+
+        @Override
+        public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script) {
+            // TODO: Implement this or get rid of ANY
+            throw new UnsupportedOperationException("ValuesSourceType.ANY is still a special case");
+        }
+
+        @Override
+        public ValuesSource replaceMissing(ValuesSource valuesSource, Object rawMissing, DocValueFormat docValueFormat, LongSupplier now) {
+            return BYTES.replaceMissing(valuesSource, rawMissing, docValueFormat, now);
+        }
+    },
+    NUMERIC {
+        @Override
+        public ValuesSource getEmpty() {
+            return ValuesSource.Numeric.EMPTY;
+        }
+
+        @Override
+        public ValuesSource getScript(AggregationScript.LeafFactory script, ValueType scriptValueType) {
+            return new ValuesSource.Numeric.Script(script, scriptValueType);
+        }
+
+        @Override
+        public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script) {
+
+            if (!(fieldContext.indexFieldData() instanceof IndexNumericFieldData)) {
+                // TODO: Is this the correct exception type here?
+                throw new IllegalArgumentException("Expected numeric type on field [" + fieldContext.field() +
+                    "], but got [" + fieldContext.fieldType().typeName() + "]");
+            }
+
+            ValuesSource.Numeric dataSource = new ValuesSource.Numeric.FieldData((IndexNumericFieldData)fieldContext.indexFieldData());
+            if (script != null) {
+                // What's the difference between ValuesSource.Numeric.Script and ValuesSource.Numeric.WithScript?
```

Comment:
`ValuesSource.Numeric.Script` == a pure script with no field context

`ValuesSource.Numeric.WithScript` == a field context that also has a script.  Known as a "value script", it allows the script to access `_value` special parameter which is the field's value... sorta like a simplified script because you don't have to use the `doc[]` notation.  It _might_ be used elsewhere, not sure, but at least for "value scripts"

I suggested removing these at one point (https://github.com/elastic/elasticsearch/issues/31532) but it was decided to keep them.  Might want to revisit at some point during the refactor to help simplify?

Reply:
So what you have here (`getScript()` using `Script` and `getField()` using `WithScript` if there is a script) is correct / matches the old way.

Reply:
Cool, thank you for that context.  I'll add some javadoc to note that for the next time we're poking around in here ;)